pmlc/dialect/pxa/transforms/subgroups.cc

```diff
@@ -159,58 +175,153 @@ struct SubgroupCostModel {
+                  SmallVector<int64_t, 4> tensorStrides) {
+    // Start with one cache line
+    double cacheLines = 1.0;
+    // Current accumulated maximum value
+    int64_t maxVal = 0;
+    // For each dimension (in sorted order)
+    assert(tensorDimensions.size() == tensorStrides.size());
+    for (size_t i = tensorDimensions.size(); i > 0; i--) {
+      // Compute gap per step
+      int64_t gap = std::abs(tensorStrides[i - 1]) - maxVal;
+      // Multiply current cache hits by size
+      cacheLines *= static_cast<double>(tensorDimensions[i - 1]);
+      // Compute probability that cache line is shared across gap
+      double probShared = 0.0; // Assume it's never shared
+      if (cacheElems != 0.0 && gap < cacheElems) {
+        probShared = 1.0 - (gap / cacheElems);
+      }
+      // Subtract shared elements
+      cacheLines -=
+          probShared * static_cast<double>(tensorDimensions[i - 1] - 1);
+      // Update maxVal
+      maxVal += std::abs(tensorStrides[i - 1]) * (tensorDimensions[i - 1] - 1);
+    }
+    return cacheLines;
+  }
+
   struct MemInfo {
     unsigned subgroupCount;
-    int64_t memSize;
+    int64_t registers;
+    int64_t accesses;
+    double cacheMiss;
+    SmallVector<int64_t, 4> tensorStrides;
+    SmallVector<int64_t, 4> tensorDimensions;
   };
 
-  // Compute the memory info for a single load/store given the current plan
-  MemInfo computeMemoryInfo(StrideInfo si) {
-    MemInfo out = {0, 1};
-    for (unsigned i = 0; i < ranges.size(); i++) {
+  size_t getIndexByArgument(mlir::BlockArgument arg) {
+    size_t index = ranges.size();
+    for (size_t i = 0; i < ranges.size(); ++i) {
       auto iv = op.getIVs()[i];
-      if (si.strides.count(iv)) {
-        if (plan.subgroupTile[i] == plan.subgroupSize && si.strides[iv] == 1) {
-          out.subgroupCount++;
-        }
-        out.memSize *= plan.innerTile[i];
+      if (iv == arg) {
+        index = i;
+        break;
       }
     }
-    if (out.subgroupCount) {
-      out.memSize /= plan.subgroupSize;
+    assert(index != ranges.size());
+    return index;
+  }
+
+  MemInfo computeMemoryInfo(SmallVector<StrideInfo, 4> dimStrides,
+                            StrideInfo flatStrides) {
+    MemInfo out = {0, 1, 1, 0.0};
+    for (auto positionalDimStrides : dimStrides) {
+      // no strides, nothing to do
+      if (positionalDimStrides.strides.empty()) {
+        continue;
+      }
+
+      int64_t pos = 0;
+      int64_t neg = 0;
+      int64_t tensorStride = 0;
+      bool hasTensorStride = false;
+
+      for (auto kvp : positionalDimStrides.strides) {
+        auto index = getIndexByArgument(kvp.first);
+        auto dimStride = kvp.second;
+
+        // dim strides should be a superset of flat strides
+        assert(flatStrides.strides.count(kvp.first));
+        auto flatStride = flatStrides.strides[kvp.first];
+
+        // set tensor stride the first time through the loop
+        if (!hasTensorStride) {
```

Comment:
We can get tensor stride at precompute time (when we get dimStrides + flatStrides) via a call to getStridesAndOffset (see Stride.cc flatten for an example).

Reply:
Done.